File: lint_parens.py
import sys
import glob

def check_parens(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()
    
    stack = 0
    errors = []
    
    for i, line in enumerate(lines):
        # Ignore string contents
        cleaned_line = ""
        in_string = False
        escape = False
        for char in line:
            if escape:
                escape = False
                continue
            if char == '\\':
                escape = True
                continue
            if char == '"':
                in_string = not in_string
                continue
            if not in_string and char != ';': # rudimentary comment check: ; is comment? check lexer
                 if char == ';': break # assume ; starts comment
                 cleaned_line += char

        for char in cleaned_line:
            if char == '(':
                stack += 1
            elif char == ')':
                stack -= 1
                if stack < 0:
                    errors.append(f"Line {i+1}: Extra closing parenthesis")
                    stack = 0 # reset to continue finding others
    
    if stack > 0:
        errors.append(f"End of file: Missing {stack} closing parenthesis/parentheses")
    
    if errors:
        print(f"Errors in {filename}:")
        for e in errors:
            print(f"  {e}")
        return False
    else:
        print(f"OK: {filename}")
        return True

files = glob.glob('examples/*.iris')
dirty = False
for f in files:
    if not check_parens(f):
        dirty = True

if dirty:
    sys.exit(1)

File: lint_parens.ts

import * as fs from 'fs';
import * as path from 'path';

function checkParens(file: string): boolean {
    const content = fs.readFileSync(file, 'utf-8');
    let balance = 0;
    const lines = content.split('\n');
    let error = false;

    // Filter out comments (lines starting with ;) 
    // And handle inline comments?
    // Simplified checker: just count ( and ) ignoring strings and comments

    // We'll use a simple state machine
    let inString = false;
    let inComment = false;
    let escaped = false;

    for (let i = 0; i < content.length; i++) {
        const char = content[i];

        if (inComment) {
            if (char === '\n') inComment = false;
            continue;
        }

        if (inString) {
            if (escaped) {
                escaped = false;
                continue;
            }
            if (char === '\\') {
                escaped = true;
                continue;
            }
            if (char === '"') {
                inString = false;
            }
            continue;
        }

        if (char === ';') {
            inComment = true;
            continue;
        }

        if (char === '"') {
            inString = true;
            continue;
        }

        if (char === '(') balance++;
        if (char === ')') balance--;

        if (balance < 0) {
            console.error(`Error: Unexpected closing parenthesis in ${file} at index ${i}`);
            return false;
        }

        if (char === '\n') {
            console.log(`Line ${content.substring(0, i).split('\n').length - 1}: Balance ${balance}`);
        }
    }

    if (balance !== 0) {
        console.error(`Error: Unbalanced parentheses in ${file}. Final balance: ${balance}`);
        return false;
    }
    return true;
}

const target = process.argv[2];
if (target) {
    if (!checkParens(target)) process.exit(1);
    console.log("Parentheses balanced.");
} else {
    const dir = 'examples';
    const files = fs.readdirSync(dir).filter(f => f.endsWith('.iris'));
    let pass = true;
    for (const f of files) {
        if (!checkParens(path.join(dir, f))) {
            pass = false;
        }
    }
    if (pass) {
        console.log('All Iris files have balanced parentheses.');
    } else {
        process.exit(1);
    }
}
